Add a section describing how to implement your own status bar widgets (#803)
JetBrains/intellij-sdk-docs: topics/user_interface_components/status_bar_widgets.md

[//]: # (title: Status Bar Widgets)

<!-- Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file. -->

The IntelliJ Platform allows plugins to extend the IDE [status bar](https://www.jetbrains.com/help/idea/guided-tour-around-the-user-interface.html#status-bar) with additional custom widgets.

Status bar widgets are small UI elements that allow providing users with useful information and settings for the current file, project, IDE, etc.
For example, the status bar contains the widget showing the encoding of the current file, or the current VCS branch of the project.

> 

Reviewer: Due to the prominent presentation and limited space, they should be used only for information or settings that are relevant enough to be shown "always".

Author: Fixed.

The starting point for extending the status bar with new widgets is the
[`StatusBarWidgetFactory`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/StatusBarWidgetFactory.java)
interface, which is registered in the `com.intellij.statusBarWidgetFactory` extension point.

In case a widget provides information or functionality related to the editor files, consider extending the
[`StatusBarEditorBasedWidgetFactory`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/widget/StatusBarEditorBasedWidgetFactory.java)
class.

Each widget factory returns a new widget from `createWidget()`.
To control the disposing of a widget, implement the `disposeWidget()`, if you just want to dispose it, use `Disposer.dispose(widget)`.

> Any widget should implement the

Author: Fixed.

[`StatusBarWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/StatusBarWidget.java)
interface.

To reuse the IntelliJ Platform implementation, you can extend one of two classes:

- [`EditorBasedWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/EditorBasedWidget.java)
- [`EditorBasedStatusBarPopup`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/EditorBasedStatusBarPopup.java)

## EditorBasedWidget

`EditorBasedWidget` is the basic widget implementation.
To implement it, override `ID()` where returns the unique ID of the widget.

> This identifier may be needed to later get a widget instance.

Reviewer: get -> obtain

can we add a short description of use-case? maybe something like "to later obtain a widget instance, e.g. to update its status" or similar

Author: Fixed.

> The IntelliJ Platform provides several predefined widget appearance options:

Reviewer: "The IntelliJ Platform provides" repeats a lot

how about "Use one of the existing..."

Author: Fixed.

  Widget with only a text.

  Example:
  [PositionPanel](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/PositionPanel.java)

- `com.intellij.openapi.wm.StatusBarWidget.MultipleTextValuesPresentation`

  Widget with a text and a popup.

  Example:
  [DvcsStatusWidget](upsource:///platform/dvcs-impl/src/com/intellij/dvcs/ui/DvcsStatusWidget.java)

> Note that they can't be combined to get, for example, an icon and a text.
>
{type="note"}

To use the selected appearance, return a class that implements one of the above interfaces from `getPresentation()`.

To create a widget with custom content, it should implement the
[`CustomStatusBarWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/CustomStatusBarWidget.java)
interface.
Override `getComponent()` to return the custom widget's component to display.

Example:
[MemoryUsagePanel](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/MemoryUsagePanel.java)

## EditorBasedStatusBarPopup

`EditorBasedStatusBarPopup` is the basis for all widgets that have a popup with a list of actions.
For example, the encoding widget of the current file.

> First, implement `createComponent()`, where returns a component for display.

Reviewer: The component to display is returned from `createComponent()`.

Author: Fixed.

Each update of the widget IDE calls `updateComponent()` to update this component.
In `updateComponent()` implementation, you can describe how the widget should change depending on the current state.

> Implement `getWidgetState()` where returns the current state of the widget.

Reviewer: where -> to return

Author: Fixed.

This state will be passed to the `updateComponent()` when the widget is updated.

> The method accepts a file that's now open in the editor.

Reviewer: that's now open -> that is currently opened in the editor

Author: Fixed.

> To create your own state class, inherit it from `com.intellij.openapi.wm.impl.status.EditorBasedStatusBarPopup.WidgetState.WidgetState`.

Author: Fixed.

> Implement `ID()`, where returns the unique ID of the widget.

Reviewer: where -> and return the ...

Author: Fixed.

This identifier may be needed to later get a widget instance.

> Implement `createInstance()`, where returns the new widget instance.

Reviewer: where -> and return...

Author: Fixed.

> Finally, implement the `createPopup()` method where returns the [popup](popups.md) that will be displayed when the widget is clicked.

Author: Fixed.

Also, it can be useful to override `registerCustomListeners()` to add custom listeners.

> This can be useful if you want to set custom listeners for a widget that will update it.

Author: Fixed.

To update a widget, use `update()`.

## Controlling Widgets Programmatically

By default, when adding a widget to the status bar, it can be displayed/hidden through the context menu of the status bar or widget.

If you want to change visibility programmatically use

> `com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetSettings.setEnabled()`.

Reviewer: ```
[`StatusBarWidgetSettings.setEnabled()`](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/widget/StatusBarWidgetSettings.kt)
```

Author: Fixed.

The first argument to the method is the factory that created the widget.

> To get it, use `com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetsManager.findWidgetFactory()` where pass the widget ID and a boolean value that describes whether the widget will be visible or not.

Reviewer: where pass -> and pass

Author: Fixed.

> To get it, use `com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetsManager.findWidgetFactory()` where pass the widget ID and a boolean value that describes whether the widget will be visible or not.

Reviewer: ```
[`StatusBarWidgetsManager.findWidgetFactory()`](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/widget/StatusBarWidgetsManager.java)
```

Author: Fixed.

> Also, you need to update the widget for the changes to take effect with `com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetsManager.updateWidget()`.

Reviewer: ```
[`StatusBarWidgetsManager.updateWidget()`](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/widget/StatusBarWidgetsManager.java)
```

## Showing widget in LightEdit mode

> By default, widgets aren't shown in LightEdit mode.

Reviewer: link https://www.jetbrains.com/help/idea/lightedit-mode.html

Author: Fixed.

> To show a widget, implement the `com.intellij.ide.lightEdit.LightEditCompatible` interface in your factory.

Reviewer: ```
... implement 
[`LightEditCompatible`](upsource:///platform/core-api/src/com/intellij/ide/lightEdit/LightEditCompatible.java) in your factory.
```

Author: Fixed.

[//]: # (title: Status Bar Widgets)

<!-- Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file. -->

The IntelliJ Platform allows plugins to extend the IDE [status bar](https://www.jetbrains.com/help/idea/guided-tour-around-the-user-interface.html#status-bar) with additional custom widgets.

Status bar widgets are small UI elements that allow providing users with useful information and settings for the current file, project, IDE, etc.
For example, the status bar contains the widget showing the encoding of the current file, or the current VCS branch of the project.

> Due to the prominent presentation and limited space, they should be used only for information or settings that are relevant enough to be shown "always".

Reviewer 2: ```suggestion
Due to the prominent presentation and limited space, they should be used only for information or settings that are relevant enough to be "always" shown.
```

Author: Fixed.

[//]: # (title: Status Bar Widgets)

<!-- Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file. -->

> The status bar widget implementation is used to add new widgets to the

Reviewer 3: ```suggestion
The IntelliJ Platform allows plugins to extend the IDE [status bar](https://www.jetbrains.com/help/idea/guided-tour-around-the-user-interface.html#status-bar) with additional custom widgets.
```

[status bar](https://www.jetbrains.com/help/idea/guided-tour-around-the-user-interface.html#status-bar).

> Widgets on the status bar are small interface elements that allow you to provide the user important information about the current file or the project.

Reviewer 3: ```suggestion
Status bar widgets are small UI elements that allow providing users with useful information and settings for the current file, project, IDE, etc.
```

Not sure my version sounds good, but the point is that it provides stuff not only for files and projects.

Author: Sounds pretty good! Thank you, fixed.

> For example, the widget showing the encoding of the current file or the current VCS branch of the project.

Reviewer 3: ```suggestion
For example, the status bar contains the widget showing the encoding of the current file or the current VCS branch of the project.
```

Author: Fixed.

The starting point for extending the status bar with new widgets is the
[`StatusBarWidgetFactory`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/StatusBarWidgetFactory.java)
interface, which is registered in the `com.intellij.statusBarWidgetFactory` extension point.

> To reuse the IntelliJ Platform implementation, you can extend the

Reviewer 3: ```suggestion
In case a widget provides information or functionality related to the editor files, consider extending the
```

Author: Fixed.

[`StatusBarEditorBasedWidgetFactory`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/widget/StatusBarEditorBasedWidgetFactory.java)
class.

> > Note that widgets inherited from this class will only work when at least one file is open.

Reviewer 3: If the previous comment is accepted, I suggest removing this callout completely as it seems redundant.

> If you want your widget to work regardless of open files, implement the `StatusBarWidgetFactory` directly.
>
{type="note"}

> Each widget factory returns a new widget from `createWidget()`.

Reviewer 3: ```suggestion
Each widget factory returns a new widget from the `createWidget() method`.
```

Reviewer: I prefer "from createWidget()"

To control the disposing of a widget, implement the `disposeWidget()`, if you just want to dispose it, use `Disposer.dispose(widget)`.

Any widget should implement the
[`StatusBarWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/StatusBarWidget.java)
interface.

To reuse the IntelliJ Platform implementation, you can extend one of two classes:

> - [`EditorBasedWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/EditorBasedWidget.java).

Reviewer 3: ```suggestion
- [`EditorBasedWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/EditorBasedWidget.java)
```

Author: Fixed.

> - [`EditorBasedStatusBarPopup`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/EditorBasedStatusBarPopup.java).

Reviewer 3: ```suggestion
- [`EditorBasedStatusBarPopup`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/impl/status/EditorBasedStatusBarPopup.java)
```

Author: Fixed.

> ## `EditorBasedWidget`

Reviewer 3: ```suggestion
## EditorBasedWidget
```
Inline code doesn't work with our rendering engine.

  [PowerSaveStatusWidgetFactory](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/PowerSaveStatusWidgetFactory.java)

- `com.intellij.openapi.wm.StatusBarWidget.TextPresentation`

  Widget with only a text.

  Example:
  [PositionPanel](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/PositionPanel.java)

- `com.intellij.openapi.wm.StatusBarWidget.MultipleTextValuesPresentation`

  Widget with a text and a popup.

  Example:
  [DvcsStatusWidget](upsource:///platform/dvcs-impl/src/com/intellij/dvcs/ui/DvcsStatusWidget.java)

> Note that they can't be combined to get, for example, an icon and a text.
>
{type="note"}

To use the selected appearance, return a class that implements one of the above interfaces from `getPresentation()`.

To create a widget with custom content, it should implement the
[`CustomStatusBarWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/CustomStatusBarWidget.java)
interface.
Override `getComponent()` to return the custom widget's component to display.

Example:
[MemoryUsagePanel](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/MemoryUsagePanel.java)

> ## `EditorBasedStatusBarPopup`

Author: Fixed.

First, implement `createComponent()`, where returns a component for display.
Each update of the widget IDE calls `updateComponent()` to update this component.
In `updateComponent()` implementation, you can describe how the widget should change depending on the current state.

Implement `getWidgetState()` where returns the current state of the widget.
This state will be passed to the `updateComponent()` when the widget is updated.
The method accepts a file that's now open in the editor.
To create your own state class, inherit it from `com.intellij.openapi.wm.impl.status.EditorBasedStatusBarPopup.WidgetState.WidgetState`.

Implement `ID()`, where returns the unique ID of the widget.
This identifier may be needed to later get a widget instance.

Implement `createInstance()`, where returns the new widget instance.

Finally, implement the `createPopup()` method where returns the [popup](popups.md) that will be displayed when the widget is clicked.

Also, it can be useful to override `registerCustomListeners()` to add custom listeners.
This can be useful if you want to set custom listeners for a widget that will update it.

To update a widget, use `update()`.

> ## How to change visibility programmatically

Reviewer 3: I suggest changing questions to regular titles as this is not an FAQ page.
Example:
"Controlling Widgets Programmatically" or similar and I would merge the next "How to update widget programmatically" to this new section.

By default, when adding a widget to the status bar, it can be displayed/hidden through the context menu of the status bar or widget.

If you want to change visibility programmatically use
`com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetSettings.setEnabled()`.

The first argument to the method is the factory that created the widget.
To get it, use `com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetsManager.findWidgetFactory()` where pass the widget ID and a boolean value that describes whether the widget will be visible or not.

Also, you need to [update](#how-to-update-widget-programmatically) the widget for the changes to take effect.

## How to update widget programmatically

To update a widget, use `com.intellij.openapi.wm.impl.status.widget.StatusBarWidgetsManager.updateWidget()`.

> ## How to show widget in LightEdit mode

Reviewer 3: ```suggestion
## Showing widget in LightEdit mode
```

Author: Fixed.
